instrumentation/jmx-metrics/library/src/main/java/io/opentelemetry/instrumentation/jmx/yaml/JmxRule.java

```diff
@@ -162,44 +164,85 @@ public MetricDef buildMetricDef() throws Exception {
-      metricExtractors[n++] = metricExtractor;
+        // generate one metric extractor per state metric key
+        // each metric extractor will have the state attribute replaced with a constant
+        for (String key : stateMapping.getStateKeys()) {
+          List<MetricAttribute> stateMetricAttributes =
+              attributeList.stream()
+                  .map(
+                      ma -> {
+                        if (!ma.isStateAttribute()) {
+                          return ma;
+                        } else {
+                          return new MetricAttribute(
+                              ma.getAttributeName(), MetricAttributeExtractor.fromConstant(key));
+                        }
+                      })
+                  .collect(Collectors.toList());
+
+          BeanAttributeExtractor stateMetricExtractor =
+              new BeanAttributeExtractor(attrExtractor.getAttributeName()) {
+
+                @Override
+                protected Number extractNumericalAttribute(
+                    MBeanServerConnection connection, ObjectName objectName) {
+                  String rawStateValue = attrExtractor.extractValue(connection, objectName);
+                  String mappedStateValue = stateMapping.getStateValue(rawStateValue);
+                  return key.equals(mappedStateValue) ? 1 : 0;
+                }
+              };
+
+          metricExtractors.add(
+              new MetricExtractor(
+                  stateMetricExtractor,
+                  metricInfo,
+                  stateMetricAttributes.toArray(new MetricAttribute[0])));
+        }
+      }
     }
 
-    return new MetricDef(group, metricExtractors);
+    return new MetricDef(group, metricExtractors.toArray(new MetricExtractor[0]));
   }
 
   private static List<MetricAttribute> combineMetricAttributes(
       List<MetricAttribute> ownAttributes, List<MetricAttribute> metricAttributes) {
+
     Map<String, MetricAttribute> set = new HashMap<>();
-    for (MetricAttribute ownAttribute : ownAttributes) {
-      set.put(ownAttribute.getAttributeName(), ownAttribute);
+    if (ownAttributes != null) {
+      for (MetricAttribute ownAttribute : ownAttributes) {
+        set.put(ownAttribute.getAttributeName(), ownAttribute);
+      }
     }
-
-    // Let the metric level defined attributes override own attributes
-    for (MetricAttribute metricAttribute : metricAttributes) {
-      set.put(metricAttribute.getAttributeName(), metricAttribute);
+    if (metricAttributes != null) {
+      // Let the metric level defined attributes override own attributes
+      for (MetricAttribute metricAttribute : metricAttributes) {
+        set.put(metricAttribute.getAttributeName(), metricAttribute);
+      }
     }
 
-    return new ArrayList<MetricAttribute>(set.values());
+    return new ArrayList<>(set.values());
+  }
+
+  private static StateMapping getEffectiveStateMapping(Metric m, JmxRule rule) {
```

Comment:
[for reviewer] for state mapping, the resulting mapping is not a combination of rule level and metric level definitions. Only one of them is used.